File: frontend/src/services/user-management.ts
import { authenticatedFetch } from "./auth";
import type { AgentCommand } from "../domain/inventory";
import type { ProductUser, SubscriptionAccessResponse } from "../domain/subscriptions";

export type UserOperation = "edit" | "remove";
export interface UserSettings {
  display_name: string;
  email: string | null;
  remark: string;
  is_active: boolean;
}
export interface UserManagementRead {
  user: ProductUser;
  revision: string;
  credential_count: number;
  blockers: string[];
  warnings: string[];
  access: SubscriptionAccessResponse;
}
export interface UserManagementResult extends UserManagementRead {
  commands: AgentCommand[];
}
export interface UserRemoval {
  id: string;
  username: string;
  status: "pending" | "failed" | "completed";
  requested_at: string;
  completed_at: string | null;
  servers: SubscriptionAccessResponse["servers"];
  warnings: string[];
  commands: AgentCommand[];
}

export function userSettings(user: ProductUser): UserSettings {
  return { display_name: user.display_name, email: user.email ?? null, remark: user.remark ?? "", is_active: user.is_active };
}
const base = import.meta.env.VITE_API_BASE_URL ?? "";
async function request<T>(url: string, init?: RequestInit, fetcher = authenticatedFetch): Promise<T> {
  const response = await fetcher(`${base}/api/v1${url}`, init);
  if (!response.ok) {
    const data = await response.json().catch(() => null);
    const detail = data?.detail;
    const message = typeof detail === "string" ? detail : Array.isArray(detail)
      ? detail.map((item: { loc?: unknown[]; msg?: string }) => `${item.loc?.slice(1).join(".") ?? ""}: ${item.msg ?? "Invalid value"}`).join("; ") : "";
    throw new Error(message || `User request failed (${response.status})`);
  }
  return response.json() as Promise<T>;
}
export function getUserManagement(username: string, fetcher = authenticatedFetch) {
  return request<UserManagementRead>(`/users/${encodeURIComponent(username)}/settings`, undefined, fetcher);
}
export function saveUser(username: string, settings: UserSettings, revision: string, fetcher = authenticatedFetch) {
  return request<UserManagementResult>(`/users/${encodeURIComponent(username)}/settings`, {
    method: "PUT", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ...settings, expected_revision: revision, acknowledge_runtime_restart: true }),
  }, fetcher);
}
export function removeUser(username: string, revision: string, name: string, unmanaged: boolean, fetcher = authenticatedFetch) {
  return request<UserRemoval>(`/users/${encodeURIComponent(username)}/remove`, {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ expected_revision: revision, confirm_name: name, acknowledge_runtime_restart: true, acknowledge_unmanaged_credentials: unmanaged }),
  }, fetcher);
}
export function getUserRemoval(id: string, fetcher = authenticatedFetch) {
  return request<UserRemoval>(`/user-removals/${encodeURIComponent(id)}`, undefined, fetcher);
}
export function retryUserRemoval(id: string, fetcher = authenticatedFetch) {
  return request<UserRemoval>(`/user-removals/${encodeURIComponent(id)}/retry`, { method: "POST" }, fetcher);
}

File: backend/app/open_node/domain/user_management.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from open_node.domain.inventory import AgentCommandRead
from open_node.domain.subscriptions import (
    ProductUserRead,
    SubscriptionAccessResponse,
    SubscriptionAccessServerRead,
    _strip_required_text,
)
from open_node.domain.user_limits import UserLimitOverrides, UserLimitsRead


class UserUpdate(BaseModel):
    model_config = {"extra": "forbid"}

    display_name: str = Field(min_length=1, max_length=120)
    email: str | None = Field(default=None, max_length=255)
    remark: str = Field(default="", max_length=1000)
    is_active: bool = Field(strict=True)
    limit_overrides: UserLimitOverrides | None = None
    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    acknowledge_runtime_restart: Literal[True]

    @field_validator("display_name")
    @classmethod
    def name(cls, value):
        return _strip_required_text(value, "display name")

    @field_validator("email")
    @classmethod
    def contact(cls, value):
        return _strip_required_text(value, "email") if value and value.strip() else None


class UserRemoval(BaseModel):
    model_config = {"extra": "forbid"}

    expected_revision: str = Field(pattern=r"^[a-f0-9]{64}$")
    confirm_name: str = Field(min_length=1, max_length=80)
    acknowledge_runtime_restart: Literal[True]
    acknowledge_unmanaged_credentials: bool = Field(default=False, strict=True)


class UserManagementRead(BaseModel):
    user: ProductUserRead
    revision: str
    credential_count: int
    blockers: list[str]
    warnings: list[str]
    access: SubscriptionAccessResponse
    limits: UserLimitsRead
    license_required: Literal[False] = False


class UserManagementResult(UserManagementRead):
    commands: list[AgentCommandRead]


class UserRemovalRead(BaseModel):
    id: UUID
    username: str
    status: Literal["pending", "failed", "completed"]
    requested_at: datetime
    completed_at: datetime | None = None
    servers: list[SubscriptionAccessServerRead]
    warnings: list[str]
    commands: list[AgentCommandRead] = Field(default_factory=list)
    license_required: Literal[False] = False
